### PR TITLE
Stringify Bytes

### PR DESCRIPTION
**Pull request by R4PaSs**
This PR adds some String-like functions to Bytes, along with an abstraction for Byte-based patterns.

Also, a bug was found within `Bytes::is_empty` and `FlatText` is now public since there was no real rationale to keep it private

### CODE REVIEW

File: lib/core/bytes.nit [hunk 1]
```diff
@@ -118,6 +186,69 @@ class Bytes
...
 
+	# Returns a copy of `self`
+	fun clone: Bytes do
+		var b = new Bytes.with_capacity(length)
+		b.append(self)
+		return b
+	end
+
+	# Trims off the whitespaces at the beginning and the end of `self`
+	#
```
Comment (privat) [inline lib/core/bytes.nit]: can you define whitespace (it can be ambiguous)

File: lib/core/bytes.nit [hunk 2]
```diff
@@ -118,6 +186,69 @@ class Bytes
...
+	end
+
+	# Returns a subset of the content of `self` starting at `from` and of length `count`
+	#
+	#     var b = "abcd".to_bytes
+	#     assert b.substring(1, 2).hexdigest == "6263"
+	#     assert b.substring(-1, 2).hexdigest == "61"
+	#     assert b.substring(1, 0).hexdigest == ""
+	#     assert b.substring(2, 5).hexdigest == "6364"
+	fun substring(from, count: Int): Bytes do
```
Comment (privat) [inline lib/core/bytes.nit]: is `substring` the right name?

**Comment by privat** [comment]
There is a lot of things (and lot of clones) but I have no better proposal, so +1 but I will wait for other reviewers.

File: lib/core/bytes.nit (hunk 1 above)
Comment (R4PaSs) [inline lib/core/bytes.nit]: "# NOTE: A whitespace is defined here as a byte whose value is <= 0x20" added

**Comment by xymus** [comment]
I'm still not convinced that replicating the String API on bytes is a good idea.

My two cents is that there should be a default transformation between `Bytes` and `String` so they share an API and are compatible. For example, `Bytes` could be a `Pattern` and could implement `search(p: Patterrn)`, using double dispatch we could search bytes within bytes, strings within strings and convert strings to UTF8 bytes in other cases.

But you're the expert and this will be useful so +1.

**Pull request merged**